### PR TITLE
Added caching hasher

### PR DESCRIPTION
**Pull request by Claudenw**
Fix for COLLECTIONS-750

### CODE REVIEW

**Comment by Claudenw** [comment]
I seem to have an issue with far to many commits showing up in this pull request (only the last 2 are real) can someone point me to how to get rid of the other 99?

**Comment by aherbert** [comment]
You need to update your fork of the apache repo from the current master, then branch from your forked master, do your new stuff and then PR from the recently forked branch.

Assuming you are in your forked repo and origin points to github.com/claudenw/commons-collections/
```
git checkout master
git remote add upstream git://github.com/apache/commons-collections/
git fetch upstream
git pull upstream master
git push
```
That should sync your fork with apache master. Then branch a topic and work as usual:
```
git checkout -b topic-branch-name
```
When you PR from the topic branch it should only have the few recent commits you made after diverging from master.

There are other ways to do it but keeping a fork up to date from which you branch has a lot of examples on the web.

The alternative is to clone apache's git repo and always leave it on master. You just update it by using git pull. Then you set upstream as your git fork. To do work you create a branch and set the tracking branch on your upstream:

Assuming origin points to github.com/apache/commons-collections/
```
git pull
git remote add upstream git://github.com/claudenw/commons-collections/
git checkout -b topic-branch-name
git push --set-upstream upstream topic-branch-name
```
Then all commits will go to your github fork.

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+     */
+    private final HashFunctionIdentity functionIdentity;
+
+    /**
+     * Constructs a CachingHasher from a list of arrays of hash values.
+     * <p>
+     * The list of hash values comprises a @code{List&lt;long[]&gt;} where each @code{long[]}
+     * is comprises two (2) values that are the result of hashing the original buffer.  Thus a
+     * CachingHasher that was built from five (5) buffers will have five arrays of two @code{longs} each.
+     *
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: Close HTML tags.

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+     * @throws IllegalArgumentException if the name does not indicate a cyclic
+     *                                  hashing function.
+     */
+    public CachingHasher(HashFunctionIdentity functionIdentity, long[][] buffers) {
+        this.functionIdentity = checkIdentity(functionIdentity);
+        this.buffers = Arrays.asList(buffers);
+    }
+
+    /**
+     * Check that the name is valid for this hasher.
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: "Check" -> "Checks".

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+import java.util.NoSuchElementException;
+import java.util.PrimitiveIterator;
+
+import org.apache.commons.collections4.bloomfilter.hasher.HashFunctionIdentity.ProcessType;
+
+/**
+ * An implementation of Hasher that attempts to leak as little data as possible.
+ * Each item in the hasher is represented by two (2) longs.  So this Hasher will
+ * still indicate how many items are in the hasher but will not leak the buffers
+ * that are being hashed as the @code DynamicHasher} does.
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: Correct the {@code ...} tag

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+    /**
+     * The hash function identity
+     */
+    private final HashFunctionIdentity functionIdentity;
+
+    /**
+     * Constructs a CachingHasher from a list of arrays of hash values.
+     * <p>
+     * The list of hash values comprises a @code{List&lt;long[]&gt;} where each @code{long[]}
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: {@code List<long[]>} and {@code long[]}

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+     * is comprises two (2) values that are the result of hashing the original buffer.  Thus a
+     * CachingHasher that was built from five (5) buffers will have five arrays of two @code{longs}
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: {@code long}s

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    public CachingHasher(HashFunctionIdentity functionIdentity, List<long[]> buffers) {
+        this.functionIdentity = checkIdentity(functionIdentity);
+        this.buffers = new ArrayList<long[]>(buffers);
+    }
+
+    /**
+     * Constructs a CachingHasher from an array of arrys of hash values.
+     *
+     * @param functionIdentity The identity of the function.
+     * @param buffers          an array of @code{long} arrays comprising two (2) values.
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: I don't think the (2) adds anything.

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    @Override
+    public PrimitiveIterator.OfInt getBits(Shape shape) {
+        HashFunctionValidator.checkAreEqual(getHashFunctionIdentity(),
+            shape.getHashFunctionIdentity());
+        return new IntIterator(shape);
+    }
+
+    /**
+     * Gets the long representations of the buffers.
+     * <p>
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: A bit more doc would help here, i.e. the List<long[]> will contains zero or more entries, each entry is a non-null array of length two containing the 64-bit pair output from the cyclic hash function.

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        }
+
+        /**
+         * Builds the hasher.
+         *
+         * @return A CachingHashers with the specified name, function and buffers.
+         */
+        @Override
+        public CachingHasher build() throws IllegalArgumentException {
+            List<long[]> cache = new ArrayList<long[]>();
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: use final

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        public final Builder with(byte[] property) {
+            buffers.add(property);
+            return this;
+        }
+
+        @Override
+        public final Builder with(String property) {
+            return with(property.getBytes(StandardCharsets.UTF_8));
+        }
+
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: Remove empty line

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/HashFunctionIdentityImpl.java [hunk 10]
```diff
@@ -51,7 +51,8 @@ public HashFunctionIdentityImpl(final HashFunctionIdentity identity) {
      * @param process the processes of the hash function.
      * @param signature the signature for the hash function.
      */
-    public HashFunctionIdentityImpl(final String provider, final String name, final Signedness signedness, final ProcessType process,
+    public HashFunctionIdentityImpl(final String provider, final String name, final Signedness signedness,
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/HashFunctionIdentityImpl.java]: No need for the line wrap here. It is unrelated to the PR.

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/HashFunctionValidator.java [hunk 11]
```diff
@@ -20,7 +20,7 @@
 /**
  * Contains validation for hash functions.
  */
-final class HashFunctionValidator {
+public final class HashFunctionValidator {
```
Comment (aherbert) [inline src/main/java/org/apache/commons/collections4/bloomfilter/hasher/HashFunctionValidator.java]: Why make this public here? That would fix it in the API. Currently it is only used internally.

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherBuilderTest.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+        assertTrue(iter.hasNext());
+        assertEquals(expected, iter.nextInt());
+        assertFalse(iter.hasNext());
+    }
+
+    /**
+     * Sets up the builder for testing.
+     */
+    @Before
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherBuilderTest.java]: Move this to the top of the test

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        builder = new CachingHasher.Builder(new MD5Cyclic());
+        shape = new Shape(new MD5Cyclic(), 3, 72, 17);
+    }
+
+    /**
+     * Tests that the expected bits are returned from hashing.
+     */
+    @Test
+    public void testGetBits() {
+
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: Empty line

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+        } catch (final NoSuchElementException expected) {
+            // do nothing
+        }
+
+        assertFalse(builder.with("Hello").build().isEmpty());
+    }
+
+    @Test
+    public void testNonCyclicHashFunction() {
+        HashFunctionIdentity hfi = new HashFunctionIdentity() {
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: Cleaner to use HashFunctionIdentityImpl here.

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+import org.junit.Test;
+
+/**
+ * Tests the {@link CachingHasher}.
+ */
+public class CachingHasherTest {
+    private CachingHasher.Builder builder;
+    private Shape shape;
+
+    private final HashFunctionIdentity testFunction = new HashFunctionIdentity() {
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: Cleaner to use HashFunctionIdentityImpl here.

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+
+        List<long[]> buffers = hasher.getBuffers();
+        HashFunctionIdentity id = hasher.getHashFunctionIdentity();
+
+        CachingHasher hasher2 = new CachingHasher( id, buffers );
+
+        PrimitiveIterator.OfInt iter1 = hasher.getBits(shape);
+        PrimitiveIterator.OfInt iter2 = hasher2.getBits(shape);
+
+        while (iter1.hasNext())
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: This should have failed checkstyle. Have you based on the current master?

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+        HashFunctionIdentity id = hasher.getHashFunctionIdentity();
+
+
+        CachingHasher hasher2 = new CachingHasher( id, ary );
+
+        PrimitiveIterator.OfInt iter1 = hasher.getBits(shape);
+        PrimitiveIterator.OfInt iter2 = hasher2.getBits(shape);
+
+        while (iter1.hasNext())
+        {
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: Another `{` on the wrong line

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+            // do nothing
+        }
+    }
+
+    /**
+     * Tests that retrieving bits for the wrong shape throws an exception.
+     */
+    @Test
+    public void testGetBits_WongShape() {
+
```
Comment (aherbert) [inline src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: empty line

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/HashFunctionValidator.java (hunk 11 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/hasher/HashFunctionValidator.java]: reversed - unintended for this pull

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/HashFunctionIdentityImpl.java (hunk 10 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/hasher/HashFunctionIdentityImpl.java]: reversed

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java (hunk 3 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java (hunk 4 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java (hunk 6 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java (hunk 7 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java (hunk 8 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: fixed

File: src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java (hunk 9 above)
Comment (Claudenw) [reply src/main/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasher.java]: fixed

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherBuilderTest.java (hunk 12 above)
Comment (Claudenw) [reply src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherBuilderTest.java]: fixed

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java (hunk 14 above)
Comment (Claudenw) [reply src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: fixed

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java (hunk 15 above)
Comment (Claudenw) [reply src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: fixed

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java (hunk 16 above)
Comment (Claudenw) [reply src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: fixed

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java (hunk 17 above)
Comment (Claudenw) [reply src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: fixed

File: src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java (hunk 18 above)
Comment (Claudenw) [reply src/test/java/org/apache/commons/collections4/bloomfilter/hasher/CachingHasherTest.java]: fixed

**Comment by Claudenw** [comment]
> Needs a matching unit test class ;-)

done

**Comment by aherbert** [comment]
This class makes a big assumption on how a CYCLIC HashFunction produces long output.

If the function does not increment the first long output from the hash function (with seed = 0) using a fixed increment to produce all the other values then this CachingHasher will create different indexes from a DynamicHasher using the same CYCLIC hash function.

For instance if the CYCLIC function creates further output using multiplications and rotations.

Consequently you can add this CachingHasher to a BloomFilter that uses the same HashFunctionIdentity but the CachingHasher output will be inconsistent with items that may have already been added with a different hasher using the same HashFunction.

Either:

- You test the cyclic hash function by calling it once to get the increment (as is currently done) and then again to test if your prediction for the next output is correct based on the assumption of a fixed increment.
- You make it clear in the javadoc that this function can output different indexes than a DynamicHasher when the cyclic function is not using a fixed increment.

This is an example where the removal of the hashing functionality from within the BloomFilter can lead to a comprised internal state due to lack of encapsulation.

If a cyclic hash function is mandated to use a fixed increment to generate subsequent hash values then this is not a problem. But it should be added to the ProcessType javadoc.

I was interpreting the ProcessType as either you use the entire byte[] representation of the object for each seeded hash (iterative) or you do not (cyclic). There is nothing in there to indicate it uses summation. It currently states:
```
Subsequent calls with a non-zero seed use the state to generate a new value.
```

**Comment by aherbert** [comment]
A second way to fix this is to update the CachingHasher to generate its own HashFunctionIdentity and not use the one of the HashFunction.

This should be combined with the HashFunctionIdentity.getSignature() method being changed to call the HashFunction.apply(byte[], int) method multiple times to simulate iteration of the HashFunction as a generator. The signature for example could be the result of the XOR of the first 3 outputs from the apply method with seeds 0, 1, 2.

Thus a BloomFilter can then check an input Hasher has the same signature and this removes the requirement to test equality as a match of name, process type and signedness. The output from the hashing function is the defacto signature of the hash function.

**Pull request closed without merging**